### PR TITLE
feat: add packet_loss toxic for simulating flaky network conditions

### PR DESCRIPTION
**Pull request by maskedmaxx**
## Overview
Adds the `packet_loss` toxic to toxiproxy, filling a gap for simulating flaky network conditions like poor Wi-Fi, mobile networks, and satellite links.

## What's included
- **packet_loss.go**: Core implementation using a simplified Gilbert-Elliott burst model
- **packet_loss_test.go**: Comprehensive test suite covering edge cases
- **README.md**: Documentation with attributes and examples

## Features
- `loss_rate`: Probability [0.0–1.0] that a chunk is dropped (default: 0.1)
- `correlation`: Extra drop probability when previous chunk was dropped, modeling burst loss (default: 0.0)
- Per-connection RNG to avoid contention between concurrent connections
- Automatic registration via `init()`

## Testing
All tests pass locally:
```bash
$ go test -v ./toxics/ -run TestPacketLoss
=== RUN   TestPacketLossToxicNoLoss
--- PASS: TestPacketLossToxicNoLoss (0.00s)
=== RUN   TestPacketLossToxicFullLoss
--- PASS: TestPacketLossToxicFullLoss (0.00s)
=== RUN   TestPacketLossToxicApproximateRate
--- PASS: TestPacketLossToxicApproximateRate (0.00s)
=== RUN   TestPacketLossToxicIndependentConnections
--- PASS: TestPacketLossToxicIndependentConnections (0.00s)
=== RUN   TestPacketLossToxicPipeDrains
--- PASS: TestPacketLossToxicPipeDrains (0.00s)
PASS
ok      github.com/Shopify/toxiproxy/v2/toxics  1.065s
```

## Usage examples
```bash
$ toxiproxy-cli create -l localhost:26379 -u localhost:6379 myproxy
$ toxiproxy-cli toxic add -t packet_loss -a loss_rate=0.1 -a correlation=0.3 myproxy

### CODE REVIEW

**Comment by maskedmaxx** [comment]
I have signed the CLA!

**Comment by maskedmaxx** [comment]
CLA check failing due to IP allow list restriction on my side. Let me know if you need anything from me.

**Comment by maskedmaxx** [comment]
Hey @abecevello , would you be able to take a look at this when you get a chance? It's been open a couple weeks without any reviews yet. happy to address any feedback. thanks!

**Review by abecevello — APPROVED** [review]
Looks good! Thanks for the contribution.

File: toxics/packet_loss.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	// when the current one was dropped, modelling burst packet loss (Gilbert-
+	// Elliott model simplified). Range [0.0, 1.0]. Default 0.0.
+	Correlation float64 `json:"correlation"`
+}
+ 
+// NewState satisfies the StatefulToxic interface. toxiproxy calls this once
+// per new connection so every connection gets its own RNG and drop state.
+func (t *PacketLossToxic) NewState() interface{} {
+	return &PacketLossToxicState{
+		rng: rand.New(rand.NewSource(rand.Int63())),
```
Comment (iyaz-shaikh) [inline toxics/packet_loss.go]: You may need a `#nosec G404` comment here to avoid a linter error (and anywhere we're using rand). See [here](https://github.com/Shopify/toxiproxy/blob/6459002f42b28c1a0bc6575a41e77de7ebf65d6b/toxics/slicer.go#L43) for an example

File: toxics/packet_loss.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
+package toxics
+ 
+// PacketLossToxic randomly drops StreamChunks passing through the proxy,
+// simulating packet loss / flaky network conditions.
+//
+// Attributes:
+//   - loss_rate : float64  probability [0.0–1.0] that a chunk is dropped
+//                          (default 0.1 -> 10 %)
```
Comment (iyaz-shaikh) [inline toxics/packet_loss.go]: Where are we defaulting to 0.1? I didn't see where that was happening

File: toxics/packet_loss.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+ 
+	for {
+		select {
+		case <-stub.Interrupt:
+			// toxiproxy is removing or reconfiguring this toxic; drain cleanly.
+			return
+ 
+		case chunk, ok := <-stub.Input:
+			if !ok {
+				// Upstream closed the connection.
```
Comment (iyaz-shaikh) [inline toxics/packet_loss.go:56]: Do we need a `stub.Close()` here like other [toxics](https://github.com/Shopify/toxiproxy/blob/6459002f42b28c1a0bc6575a41e77de7ebf65d6b/toxics/slicer.go#L58) do when they return?

File: toxics/packet_loss.go (hunk 1 above)
Comment (maskedmaxx) [reply toxics/packet_loss.go]: Good call, I’ll add `#nosec G404` like the slicer toxic does.

File: toxics/packet_loss.go (hunk 2 above)
Comment (maskedmaxx) [reply toxics/packet_loss.go]: You’re right, we aren’t defaulting to 0.1 anywhere. That was incorrect wording in the comments/docs. The Go zero value makes `loss_rate` default to 0.0, so I’ll update the docs/comments to say packet loss is disabled by default

File: toxics/packet_loss.go (hunk 3 above)
Comment (maskedmaxx) [reply toxics/packet_loss.go:56]: Yes, I think so. Other toxics call `stub.Close()` when the input stream ends, which closes the toxic output channel and marks the stub closed. I’ll update `packet_loss` to do the same before returning

**Comment by maskedmaxx** [comment]
@iyaz-shaikh thanks for the feedback! I pushed an update addressing your comments:

- added `#nosec G404` for the `math/rand` usage since this is only simulation randomness
- updated the input-close path to call `stub.Close()` like the other toxics
- corrected the `loss_rate` docs/comments to default `0.0`, since there wasn’t actually a `0.1` default in code
- ran `gofmt` and verified locally with `go test ./toxics -run PacketLoss`

Let me know if you’d like anything else changed.

**Review by iyaz-shaikh — APPROVED** [review]
LGTM, thanks for addressing the feedback 👍

**Comment by abecevello** [comment]
@maskedmaxx Looks like there is a new linting error: "toxics/packet_loss.go:33:1: The line is 128 characters long, which exceeds the maximum of 100 characters."

**Comment by maskedmaxx** [comment]
thanks for letting me know @abecevello. I shortened the `#nosec G404` comment so the line stays under the 100-character lint limit, should be good now.
